crates/data-generator/src/blocks/output.rs

```diff
@@ -0,0 +1,628 @@
+                    prop_enum_name: block_prop.name_pc.to_ident(),
+                })
+                .collect();
+
+            Self { properties }
+        }
+    }
+
+    impl ToTokens for BlockStruct {
+        fn to_tokens(&self, tokens: &mut TokenStream) {
+            tokens.extend(quote! {
+                #[derive(Debug, Default, Eq, PartialEq, Copy, Clone, Hash)]
+                pub struct Block
+            });
+
+            if self.properties.is_empty() {
+                // "pub struct LavaCauldron;"
+                tokens.append(Punct::new(';', Spacing::Alone));
+            } else {
+                let mut stream = TokenStream::new();
+                self.properties
+                    .iter()
+                    .for_each(|prop| prop.to_tokens(&mut stream));
+
+                // "pub struct Lectern { pub has_book: HasBook, }"
+                tokens.append(Group::new(Delimiter::Brace, stream));
+            }
+        }
+    }
+
+    pub struct BlockStructField {
+        pub prop_name: Ident,      // honey_level
+        pub prop_enum_name: Ident, // HoneyLevel
+    }
+
+    impl ToTokens for BlockStructField {
+        fn to_tokens(&self, tokens: &mut TokenStream) {
+            let prop_name = &self.prop_name;
+            let prop_enum_name = &self.prop_enum_name;
+
+            // "pub open: Open,"
+            tokens.extend(quote!(pub #prop_name: #prop_enum_name,))
+        }
+    }
+}
+
+mod block_convert {
+    use proc_macro2::{Delimiter, Group, Ident, Literal, TokenStream};
+    use quote::{quote, ToTokens, TokenStreamExt as _};
+
+    use super::StringExt as _;
+    use crate::blocks::model::{Block, Id, State};
+
+    #[derive(Debug)]
+    pub struct BlockConvert {
+        pub id_states: Vec<(Id, BlockStateInst)>,
+    }
+
+    impl From<&Block> for BlockConvert {
+        fn from(block: &Block) -> Self {
+            let id_states = block
+                .states
+                .iter()
+                .map(|(block_id, block_state)| (block_id.clone(), block_state.into()))
+                .collect();
+
+            Self { id_states }
+        }
+    }
+
+    impl ToTokens for BlockConvert {
+        fn to_tokens(&self, tokens: &mut TokenStream) {
+            BlockIntoU32(self).to_tokens(tokens);
+            BlockFromU32(self).to_tokens(tokens);
+        }
+    }
+
+    #[derive(Debug)]
+    pub struct BlockIntoU32<'a>(&'a BlockConvert);
+
+    impl ToTokens for BlockIntoU32<'_> {
+        fn to_tokens(&self, tokens: &mut TokenStream) {
+            let mut match_arms = TokenStream::new();
+            for (id, state) in &self.0.id_states {
+                // "LargeAmethystBud { Facing: North, Waterlogged: True, }"
+                let state_inst = state.to_token_stream();
+
+                // "LavaCauldron => 7402,"
+                let stream = quote! { #state_inst => #id, };
+                stream.to_tokens(&mut match_arms);
+            }
```

Comment:
Must've lost the change while rebasing. This should be:
```rs
for (id, state_inst) in &self.0.id_states {
    // "RedstoneLamp { Lit: Lit::True, } => 7417,"
    let match_arm = quote! { #state_inst => #id, };
    match_arm.to_tokens(&mut match_arms);
}
```
Or more generally, though I'm not in favour of this option:
```rs
let match_arms = self.0.id_states.iter().map(|(id, state)| {
    // "RedstoneLamp { Lit: Lit::True, } => 7417,"
    quote! { #state_inst => #id, }
});
```
Also, back to `BlockStructInst`, this is roughly what it should look like after inlining `impl ToTokens` for it.